### PR TITLE
[WIP] Navigation Menu: dropdown activation mode

### LINKED ISSUE #18395: Navigation Block: Allow configuring behaviour of dropdowns.
It'd be nice to be able to toggle the behaviour of dropdown menus to open on click / tap vs focus / hover. This could be a setting in the inspector controls.

### PR DESCRIPTION
**Pull request by retrofox**
## Description
This PR implements the ability to set the event which opens the dropdown menu in the front-end, adding a section in the inspector control.

Fixes https://github.com/WordPress/gutenberg/issues/18395

Notes:

* I'm not totally unconfident about the text defined on this block. Feedback is more than welcome!
* ~`onClick` should be handled also with js since it isn't possible with just CSS.~

## How has this been tested?
Once these changes are applied, you should be able to see the `Dropdown Menu Activation` block in the inspector control:

Screen Shot 2019-11-11 at 4 49 17 PM

It allows changing the behavior of how the dropdown menu shows in the front-end through of the selected event: `hover` or `click`.

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR. .

### CODE REVIEW

**Comment by retrofox** [comment]
Question: How are we going to deal when the menu has defined the activation mode with `click`, and the item has a link?

In some way, it's related to the possibility to re[nder an item with a not-anchor element](https://github.com/WordPress/gutenberg/pull/18442).

File: packages/block-library/src/navigation-menu/edit.js [hunk 1]
```diff
@@ -73,7 +74,7 @@ function NavigationMenu( {
 				return null;
 			}
 
-			return pages.map( ( { title, type, link: url, id } ) => (
+			return pages.map( ( { title, type, link: url, id, menuEventActivation } ) => (
```
Comment (getdave) [inline packages/block-library/src/navigation-menu/edit.js]: Wondering whether `menuEventActivation` would be clearer as `dropdownActivationEvent`. At the least swapping `Event` and `Activation` should happen to make the "thing" we are describing (ie: the event) is clearer (ie: `menuActivationEvent`).

File: packages/block-library/src/navigation-menu/index.php [hunk 2]
```diff
@@ -94,10 +98,29 @@ function render_block_navigation_menu( $attributes, $content, $block ) {
...
 	$html = '';
+	$level = 0;
 	foreach ( (array) $block['innerBlocks'] as $key => $menu_item ) {
+		$has_sub_items = count( (array) $menu_item['innerBlocks'] ) > 0;
 
 		$html .= '<li class="wp-block-navigation-menu-item ' . $colors['bg_css_classes'] . '"' . $colors['bg_inline_styles'] . '>' .
+			(
+				( $has_sub_items && ! $on_hover_mode )
+					? (
+						'<input
```
Comment (getdave) [inline packages/block-library/src/navigation-menu/index.php]: We'll need some ARIA roles here such as `aria-pressed` and `aria-expanded`.

File: packages/block-library/src/navigation-menu/index.php [hunk 3]
```diff
@@ -94,10 +98,29 @@ function render_block_navigation_menu( $attributes, $content, $block ) {
...
+							id="toggle-handler-' . $level . '-' . $key . '"
+							class="wp-block-navigation-menu-item__toggle-handler"
+							type="checkbox"
+						>' .
+						'<label
```
Comment (getdave) [inline packages/block-library/src/navigation-menu/index.php]: Perhaps take a look at [Bootstrap's "Dropdown" implementation](https://getbootstrap.com/docs/4.0/components/dropdowns/) for a11y guidance here.

File: packages/block-library/src/navigation-menu/inspector-control-menu-activation.js [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+					},
+					{
+						label: __('Click / Tap Event.' ),
+						value: 'onClick',
+					}
+				] }
+				selected={ selected }
+				onChange={ onChange }
+				label={ __( 'Dropdown menu activation' ) }
+				help={ __( 'Select the way of how the dropdown menu will be activated.' ) }
```
Comment (getdave) [inline packages/block-library/src/navigation-menu/inspector-control-menu-activation.js]: > Select the way of how the dropdown menu will be activated

Could be phrased as

> Choose the way dropdown menus will be activated

File: packages/block-library/src/navigation-menu/style.scss [hunk 5]
```diff
@@ -106,4 +90,42 @@
 		}
 	}
 
+	// Menu event activation.
+	&.is-activated-by-hover li {
+		&:hover,
+		&:focus-within {
+			cursor: pointer;
+			z-index: 99999;
```
Comment (getdave) [inline packages/block-library/src/navigation-menu/style.scss]: What's this super high value about? 😄

File: packages/block-library/src/navigation-menu/style.scss [hunk 6]
```diff
@@ -106,4 +90,42 @@
...
+		}
+
+		// Submenu Display
+		&:hover > ul,
+		&:focus-within > ul,
+		& ul:hover,
```
Comment (getdave) [inline packages/block-library/src/navigation-menu/style.scss]: Why are these needed? If the `ul` is within the `li` then the `:hover` on the `li` should sustain the hover effect.

File: packages/block-library/src/navigation-menu/style.scss [hunk 7]
```diff
@@ -106,4 +90,42 @@
...
+		& ul:focus {
+			visibility: visible;
+			opacity: 1;
+			display: block;
+		}
+	}
+
+	&.is-activated-by-click li {
+		input.wp-block-navigation-menu-item__toggle-handler {
+			display: none;
```
Comment (getdave) [inline packages/block-library/src/navigation-menu/style.scss]: This will make the control impossible to access for assistive tech. Better to use the `VisuallyHidden` component or a CSS solution to hide visually but retain access for assistive tech:

See also https://wordpress.slack.com/archives/C02QB2JS7/p1573637389298100

File: packages/block-library/src/navigation-menu/style.scss [hunk 8]
```diff
@@ -106,4 +90,42 @@
...
+		}
+
+		label.wp-block-navigation-menu-item__toggle-for {
```
Comment (getdave) [inline packages/block-library/src/navigation-menu/style.scss]: I'd avoid qualifying with `label` unless it is required for specificity. Otherwise we're increasing the specificity for no reason which I would tend to avoid at all costs.

Ditto for `input.` above.

**Review by getdave — CHANGES_REQUESTED** [review]
Nice work here.  A tricky problem. 

A few a11y concerns and minor CSS implementation points.

Also when testing `hover` mode with TwentyTwenty I saw this

Screen Shot 2019-11-13 at 09 35 16

**Comment by karmatosed** [comment]
I wanted to dig a little in why this is being added and how perhaps we could simplify. First, is this going to be seen by everyone or something you can extend? If everyone, I would suggest considering having a toggle and reducing a little to what is the default. Maybe by having a top settings section:

Frame 7

I do wonder though how this works with sub-menus as the parent is a link. Perhaps I am missing some background on this as it would be great to know.

**Comment by mtias** [comment]
What's the status on this one?

**Comment by jasmussen** [comment]
In https://github.com/WordPress/gutenberg/issues/22824#issuecomment-788916109 we are discussing a hamburger menu button. For this to work and be accessible, we can't use the checkbox hack, we have to have some JS output. Potentially that same JS could be used to toggle open submenus on click.

**Comment by getdave** [comment]
@jasmussen Do we still want to pursue this? Having the option to toggle based on hover/focus vs click could be useful. How do you see this working with the hamburger you're discussing in relation to mobile?

@retrofox are you likely to be working on this anytime soon?

**Comment by retrofox** [comment]
> @retrofox are you likely to be working on this anytime soon?

👋, Dave. I'm not sure whether I'll be able to look into this in the short term. I'm going to take a look when I have the chance. I think it depends on the answer to _- Do we still want to pursue this?_

**Comment by jasmussen** [comment]
> Do we still want to pursue this? Having the option to toggle based on hover/focus vs click could be useful. How do you see this working with the hamburger you're discussing in relation to mobile?

I think we want to eventually revisit this. However I also think it's more complicated than it seems. As an exapmle, if the primary menu item links to /products, but also opens a dropdown of subpages, how do you get to /products, if clicking on the menu item just opens the dropdown? Options:

- You can't open the dropdown by clicking the menu item, you have to click the chevron to the right of it. Consistent, but probably not obvious.
- You can only visit /products if you click twice, once to open, another to visit. Probably also not discoverable, or accessible.
- The main menu item becomes completely inert, and the /product link is duplicated inside. Could work well, likely needs some ability to customize the duplicate label, such as "See all", though potentially we could use the Menu Description field for this.
- The menu item becomes impossible to visit entirely. Not a good solution.
- A new menu item is created, "Dropdown menu", which serves only to open the dropdown on click.

It's solvable, we should solve it. But it's less obvious than a toggle, I would say.

**Comment by aristath** [comment]
As a sidenote, in the twentytwentyone theme we did something similar, allowing both hover & click events.
It was a bit tricky getting it right, but I feel it was the best of both worlds, as it allows most scenarios and is as accessible as a navigation can possibly be.
To accomplish that we inject a `<button>` (the caret) in menu items (see [markup for the button here](https://github.com/WordPress/twentytwentyone/blob/trunk/inc/menu-functions.php#L28-L40)), and then there's a script that takes care of the rest: https://github.com/WordPress/twentytwentyone/blob/trunk/assets/js/primary-navigation.js
I assume we can do the same here, and it's not a binary option (click or hover), in order to be accessible we need to take into account all interaction paths.
On a personal note I love click and strongly dislike hover... but the above implementation proved to be satisfying to all.

**Comment by jasmussen** [comment]
Thanks for sharing that, @aristath, I wasn't aware. Is this the behavior to which you are referring?

tt1

I.e. as you say, both hover and click at the same time? If yes, that does seem to be a variant of the first option I suggested, a split button. I do appreciate that it can become a toggle like you suggest, but I don't personally find the behavior entirely intuitive or obvious.

That said, I'd be super happy to help polish a PR that landed that exact behavior as a baseline, and then we could always go from there.

**Comment by aristath** [comment]
> Is this the behavior to which you are referring?

Yes. The implementation adds a split button, and initially we had submenus only expand on click. But there were a few complaints from people too used to the on-hover action, so the combo behavior was a compromise between the 2.
It works great with keyboard navigation though, which was a big priority.

Using a split button opens up a lot of possibilities... It can work on hover, on click or both, and styling-wise there's a lot we can do 👍
